src/main/java/gregtech/common/covers/filter/SmartItemFilter.java

```diff
@@ -12,84 +14,127 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.util.IStringSerializable;
 
+import com.cleanroommc.modularui.api.drawable.IKey;
+import com.cleanroommc.modularui.screen.ModularPanel;
+import com.cleanroommc.modularui.utils.Color;
+import com.cleanroommc.modularui.value.BoolValue;
+import com.cleanroommc.modularui.value.sync.EnumSyncValue;
+import com.cleanroommc.modularui.value.sync.GuiSyncManager;
+import com.cleanroommc.modularui.widget.Widget;
+import com.cleanroommc.modularui.widgets.ToggleButton;
+import com.cleanroommc.modularui.widgets.layout.Column;
+import com.cleanroommc.modularui.widgets.layout.Row;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
 import org.jetbrains.annotations.NotNull;
 
 import java.util.Collections;
-import java.util.Map;
 import java.util.function.Consumer;
 
 public class SmartItemFilter extends ItemFilter {
 
-    private SmartFilteringMode filteringMode = SmartFilteringMode.ELECTROLYZER;
+    private SmartItemFilterReader filterReader;
 
-    public SmartFilteringMode getFilteringMode() {
-        return filteringMode;
-    }
-
-    public void setFilteringMode(SmartFilteringMode filteringMode) {
-        this.filteringMode = filteringMode;
-        markDirty();
+    public SmartItemFilter(ItemStack stack) {
+        this.filterReader = new SmartItemFilterReader(stack);
+        setFilterReader(this.filterReader);
     }
 
-    @Override
-    public int getSlotTransferLimit(Object matchSlot, int globalTransferLimit) {
-        ItemAndMetadataAndStackSize itemAndMetadata = (ItemAndMetadataAndStackSize) matchSlot;
-        return itemAndMetadata.transferStackSize;
+    public SmartFilteringMode getFilteringMode() {
+        return this.filterReader.getFilteringMode();
     }
 
     @Override
-    public Object matchItemStack(ItemStack itemStack) {
-        ItemAndMetadata itemAndMetadata = new ItemAndMetadata(itemStack);
-        Integer cachedTransferRateValue = filteringMode.transferStackSizesCache.get(itemAndMetadata);
+    public int getTransferLimit(ItemStack stack, int globalTransferLimit) {
+        ItemAndMetadata itemAndMetadata = new ItemAndMetadata(stack);
+        var filterMode = this.filterReader.getFilteringMode();
+        int cachedTransferRateValue = filterMode.transferStackSizesCache.getOrDefault(itemAndMetadata, -1);
 
-        if (cachedTransferRateValue == null) {
-            ItemStack infinitelyBigStack = itemStack.copy();
+        if (cachedTransferRateValue == -1) {
```

Comment:
Should this be `< 0`?

Reply:
could, but the only value that should be returned that's `< 0` is `-1`